prov/tcp/src/tcpx_progress.c

```diff
@@ -144,6 +145,47 @@ static int tcpx_process_remote_write(struct tcpx_pe_entry *pe_entry)
 	return FI_SUCCESS;
 }
 
+static int tcpx_process_remote_read_req(struct tcpx_pe_entry *pe_entry)
+{
+	struct tcpx_ep *tcpx_ep = pe_entry->ep;
+	struct tcpx_cq *tcpx_cq;
+	struct tcpx_pe_entry *resp_entry;
+	int i;
+
+	tcpx_cq = container_of(tcpx_ep->util_ep.rx_cq, struct tcpx_cq,
+			       util_cq);
+
+	resp_entry = tcpx_pe_entry_alloc(tcpx_cq);
+	if (!resp_entry)
+		return -FI_EAGAIN;
+
+	resp_entry->msg_data.iov[0].iov_base = (void *) &resp_entry->msg_hdr;
+	resp_entry->msg_data.iov[0].iov_len = sizeof(resp_entry->msg_hdr);
+	resp_entry->msg_data.iov_cnt = 1 + pe_entry->msg_hdr.rma_iov_cnt;
+
+	for ( i = 0 ; i < pe_entry->msg_hdr.rma_iov_cnt ; i++ ) {
+		resp_entry->msg_data.iov[i+1].iov_base =
+			(void *) pe_entry->msg_hdr.rma_iov[i].addr;
+		resp_entry->msg_data.iov[i+1].iov_len =
+			pe_entry->msg_hdr.rma_iov[i].len;
+	}
+
+	resp_entry->msg_hdr.hdr.version = OFI_CTRL_VERSION;
+	resp_entry->msg_hdr.hdr.op = ofi_op_read_rsp;
+	resp_entry->msg_hdr.hdr.op_data = TCPX_OP_REMOTE_READ_RSP;
+	resp_entry->msg_hdr.hdr.size =
+		htonll(ofi_total_iov_len(resp_entry->msg_data.iov,
+					 resp_entry->msg_data.iov_cnt));
```

Comment:
We can calculate the total size in the for loop above, and avoid looping over the iov twice.